### PR TITLE
Reuse joins generated by Where and OrderBy methods when fetching for Linq provider

### LINKED ISSUE #871: NH-3115 - Duplicate joins when using fetching with where in LINQ query
**Andrey Kozhyn** created an issue — <time datetime="2012-04-12T13:36:37.835Z" title="2012-04-12T13:36:37.835Z">12th April 2012, 13:36:37</time>:
> This query:
> 
> ```
> 
> var orderedQueryable = this.participationRequests
>            .Fetch(x => x.CommunityEvent)
>            .Fetch(x => x.CommunityMember)
>                 .ThenFetch(x => x.User)
>            .Where(x => x.CommunityMember.Community.Id == communityId)
>            .OrderBy(x => x.CreateDate);
> 
> ```
> 
> Produces the following SQL:
> 
> ```sql
> 
> select *
> from   ParticipationRequests participat0_
>        left outer join CommunityEvents communitye1_
>          on participat0*.CommunityEventId = communitye1*.Id
>        left outer join CommunityMembers communitym2_
>          on participat0*.CommunityMemberId = communitym2*.Id
>        left outer join Users user3_
>          on communitym2*.UserId = user3*.Id
>        inner join CommunityMembers communitym4_
>          on participat0*.CommunityMemberId = communitym4*.Id
>        inner join CommunityMembers communitym5_
>          on participat0*.CommunityMemberId = communitym5*.Id
>        inner join Communities community6_
>          on communitym5*.CommunityId = community6*.Id
> where  community6_.Id = 2002 /** @p0 **/
> order  by participat0_.CreateDate asc
> 
> ```
> 
> It generates left outer join and inner join, should be only one of them
> 
> Some additional links:
> 
> http://stackoverflow.com/questions/8152298/linq-to-nhibernate-duplicates-joins
> http://stackoverflow.com/questions/10104949/nhibernate-generating-outer-join-for-a-fetch
> 
---
**Andrey Kozhyn** added a comment — <time datetime="2012-04-12T14:14:57.292Z" title="2012-04-12T14:14:57.292Z">12th April 2012, 14:14:57</time>:
> Here is a test case for this issue
> 
---
**Alexander Zaytsev** added a comment — <time datetime="2012-04-12T15:58:09.579Z" title="2012-04-12T15:58:09.579Z">12th April 2012, 15:58:09</time>:
> Could you please check with NH 3.3.0CR1? 
> 
---
**Andrey Kozhyn** added a comment — <time datetime="2012-04-13T06:53:00.905Z" title="2012-04-13T06:53:00.905Z">13th April 2012, 6:53:00</time>:
> Same in 3.3.0CR1. Added a new test case for new version
>

### LINKED ISSUE #2495: Fetch joins in LINQ should be reused by joins generated by Where statement
See https://github.com/nhibernate/nhibernate-core/pull/2494#issuecomment-678422305

### PR DESCRIPTION
**Pull request by maca88**
Fixes #871

### CODE REVIEW

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs [hunk 1]
```diff
@@ -379,6 +391,38 @@ public override void VisitResultOperator(ResultOperatorBase resultOperator, Quer
 			ResultOperatorMap.Process(resultOperator, this, _hqlTree);
 		}
 
+		private bool TryGetRelatedFetchRequest(NhJoinClause joinClause, QueryModel queryModel, out FetchRequestBase fetchRequest)
+		{
+			if (joinClause.Restrictions.Count > 0 ||
+			    !(joinClause.FromExpression is MemberExpression memberExpression) ||
+			    !(memberExpression.Expression is QuerySourceReferenceExpression querySource) ||
+				// Hql does not support fetch with group by and select
+			    queryModel.ResultOperators.Any(o => o is GroupResultOperator) ||
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:400]: Can you clarify why this check is needed? Is it some kind of optimization?

Comment (maca88) [reply src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:400]: Reusing a join for group by and select does not work, the fetch join won't be added to the query, so a normal hql join must be used. I didn't add the test for group by as it already exists in `GroupByTests`:
https://github.com/nhibernate/nhibernate-core/blob/aadc6ce23ba659abbd91ba3e6518f799d2a5e4f3/src/NHibernate.Test/Linq/ByMethod/GroupByTests.cs#L905-L910

File: src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs [hunk 2]
```diff
@@ -68,8 +68,18 @@ private void Process(FetchRequestBase resultOperator, QueryModelVisitor queryMod
 				}
 			}
 
-			var alias = queryModelVisitor.Model.GetNewName("_");
-			tree.AddFromClause(tree.TreeBuilder.LeftFetchJoin(memberPath, tree.TreeBuilder.Alias(alias)));
+			var relatedJoin = queryModelVisitor.RelatedJoinFetchRequests.FirstOrDefault(o => o.Value == resultOperator).Key;
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs:71]: Shouldn't this logic be moved to  `if (resultOperator is FetchOneRequest)` block? It seems we shouldn't try to reuse collection fetches

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs [hunk 3]
```diff
@@ -121,6 +122,9 @@ public class QueryModelVisitor : NhQueryModelVisitorBase, INhQueryModelVisitor,
 
 		public ResultOperatorRewriterResult RewrittenOperatorResult { get; private set; }
 
+		internal Dictionary<NhJoinClause, FetchRequestBase> RelatedJoinFetchRequests { get; } =
+			new Dictionary<NhJoinClause, FetchRequestBase>();
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/QueryModelVisitor.cs:126]: Same question.. Shouldn't it be restricted to `FetchOneRequest`?

File: src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs (hunk 2 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs]: It should, moved.

File: src/NHibernate/Linq/Visitors/QueryModelVisitor.cs (hunk 3 above)
Comment (maca88) [reply src/NHibernate/Linq/Visitors/QueryModelVisitor.cs]: It should, restricted.

File: src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs [hunk 4]
```diff
@@ -29,7 +29,8 @@ public void Process(FetchRequestBase resultOperator, QueryModelVisitor queryMode
 
 		private void Process(FetchRequestBase resultOperator, QueryModelVisitor queryModelVisitor, IntermediateHqlTree tree, HqlDot memberPath, IType propType)
 		{
-			if (resultOperator is FetchOneRequest)
+			string alias = null;
+			if (resultOperator is FetchOneRequest fetchRequest)
```
Comment (bahusoid) [inline src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs:33]: `fetchRequest` variable seems not really required. `resultOperator` can still be used for comparison.

Comment (maca88) [reply src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessFetch.cs]: Variable removed.

**Review by bahusoid — APPROVED** [review]

**Review by bahusoid — APPROVED** [review]

**Pull request merged**